### PR TITLE
small optimization to `f(...; kw...)` when `kw` is empty

### PR DESCRIPTION
**Pull request by JeffBezanson**
Found a modest piece of low-hanging fruit here. In this case, we were "copying" the container `kw` to a new (empty) array, and then checking whether that is empty, when we could just check whether `kw` is empty. This comes up a lot in definitions like

```
call(f, args...; kwargs...) = f(args...; kwargs...)
```

(e.g. `remotecall`), where allocating an unnecessary empty array has significant overhead for what should be a no-op. This shaves off another ~10% of the memory allocated in the benchmark in #21543.

This also takes advantage of some common cases where `kw` has a known length, where before we always copied it using `push!`.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
Ha, that failure was a good puzzle. Apparently `@testset` depended on the (unnecessary) `for` loop generated by keyword arg lowering to force toplevel expressions to be compiled, in order to get better backtraces.

File: base/test.jl [hunk 1]
```diff
@@ -848,6 +848,9 @@ function testset_beginend(args, tests)
     # action (such as reporting the results)
     quote
         ts = $(testsettype)($desc; $options...)
+        # this empty loop is here to force the block to be compiled,
+        # which is needed for backtrace scrubbing to work correctly.
+        while false; end
```
Comment (tkelman) [inline base/test.jl]: this doesn't seem fragile at all

**Comment by tkelman** [comment]
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21601#issuecomment-297978303) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/bec8bf378c983fda7aa76db696cebb4310aeb23f/c1b958b_vs_e6ab126/report.md). cc @jrevels

**Comment by Sacha0** [comment]
Chances are the potential regressions are noise. But [it's dangerous to go alone, so take this](http://knowyourmeme.com/memes/it-s-dangerous-to-go-alone-take-this) tiny metal pegasus @nanosoldier `runbenchmarks("array" || "sparse" || "tuple", vs = ":master")`

File: base/test.jl (hunk 1 above)
Comment (JeffBezanson) [inline base/test.jl]: Ok, suggest an alternative. There is a test for this at least.

Comment (TotalVerb) [inline base/test.jl]: Would putting it in a `let` work?

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21601#issuecomment-298022516) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/9bc54606565400edef57fc172d1640694fc7dd00/c1b958b_vs_e6ab126/report.md). cc @jrevels

File: base/test.jl (hunk 1 above)
Comment (vtjnash) [inline base/test.jl]: I think this should be fine. If we change the front-end heuristics, we'll worry about this then. Probably by then backtrace scrubbing will be decoupled from compilation anyways.

Comment (vtjnash) [inline base/test.jl]: But maybe put a comment on whichever test led you to notice this, as a breadcrumb for the next person who runs into this?

**Comment by Sacha0** [comment]
The potential `["tuple","reduction",("sumabs",(4,))]` and `["sparse","transpose",("ctranspose!",(20000,10000))]` regressions are consistent across nanosoldier runs, but those two benchmarks are perennially noisy so I'm not certain what to make of that consistency...

**Comment by JeffBezanson** [comment]
They're consistently inconsistent :)

**Comment by JeffBezanson** [comment]
Interesting failure on os x. Log backed up:
https://gist.github.com/JeffBezanson/638cba8d73c1b46963e4971135db70ab

**Pull request merged**